gdnative-core/src/core_types/variant_array.rs

```diff
@@ -272,163 +271,203 @@ impl VariantArray<Unique> {
     pub fn insert<T: OwnedToVariant>(&self, at: i32, val: T) {
         unsafe { (get_api().godot_array_insert)(self.sys_mut(), at, val.owned_to_variant().sys()) }
     }
+}
 
-    /// Returns an iterator through all values in the `VariantArray`.
+/// Operations allowed on non-unique arrays.
+impl<Access: NonUniqueThreadAccess> VariantArray<Access> {
+    /// Assume that this is the only reference to this array, on which
+    /// operations that change the container size can be safely performed.
     ///
-    /// `VariantArray` is reference-counted and have interior mutability in Rust parlance.
-    /// Modifying the same underlying collection while observing the safety assumptions will
-    /// not violate memory safely, but may lead to surprising behavior in the iterator.
+    /// # Safety
+    ///
+    /// It isn't thread-safe to perform operations that change the container
+    /// size from multiple threads at the same time.
+    /// Creating multiple `Unique` references to the same collections, or
+    /// violating the thread-safety guidelines in non-Rust code will cause
+    /// undefined behavior.
     #[inline]
-    pub fn iter(&self) -> IterUnique {
-        self.into_iter()
+    pub unsafe fn assume_unique(self) -> VariantArray<Unique> {
+        self.cast_access()
+    }
+}
+
+/// Operations allowed on unique arrays.
+impl VariantArray<Unique> {
+    /// Creates an empty `VariantArray`.
+    #[inline]
+    pub fn new() -> Self {
+        unsafe {
+            let mut sys = sys::godot_array::default();
+            (get_api().godot_array_new)(&mut sys);
+            Self::from_sys(sys)
+        }
+    }
+
+    /// Put this array under the "shared" access type.
+    #[inline]
+    pub fn into_shared(self) -> VariantArray<Shared> {
+        unsafe { self.cast_access() }
+    }
+
+    /// Put this array under the "thread-local" access type.
+    #[inline]
+    pub fn into_thread_local(self) -> VariantArray<ThreadLocal> {
+        unsafe { self.cast_access() }
     }
 }
 
+/// Operations allowed on arrays that might be shared between different threads.
 impl VariantArray<Shared> {
     /// Create a new shared array.
     #[inline]
     pub fn new_shared() -> Self {
         VariantArray::<Unique>::new().into_shared()
     }
 
-    /// Assume that this is the only reference to this array, on which
-    /// operations that change the container size can be safely performed.
+    /// Clears the array, resizing to 0.
     ///
-    /// # Safety
+    #[doc_variant_collection_safety]
```

Comment:
This proc-macro is added so we don't have to copy-paste the same block of text some 20 times. An attribute macro is required because ordinary macros can't expand into attributes.